internal/fxlog/fxlog.go

```diff
@@ -21,86 +21,127 @@
 package fxlog
 
 import (
+	"strings"
+
+	"go.uber.org/fx/internal/fxreflect"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
-// Level is the level of logging used by Logger.
-type Level int
+// Logger defines interface used for logging.
+type Logger interface {
+	// LogEvent is called when a logging event is emitted.
+	LogEvent(Event)
+}
 
-const (
-	// InfoLevel is Info logging level used to log messages via zap.
-	InfoLevel Level = iota
-	// ErrorLevel is Info logging level used to log messages via zap.
-	ErrorLevel
-)
+// Event defines an event emitted by fx.
+type Event interface{}
 
-// Entry is an entry to be later serialized into zap message and fields.
-type Entry struct {
-	Level   Level
-	Message string
-	Fields  []Field
-	Stack   string
+// LifecycleOnStartEvent is emitted for whenever an OnStart hook is executed
+type LifecycleOnStartEvent struct {
+	Caller string
 }
 
-// WithStack mutates an Entry to add a Stack field.
-func (e Entry) WithStack(stack string) Entry {
-	e.Stack = stack
-	return e
+// LifecycleOnStopEvent is emitted for whenever an OnStart hook is executed
+type LifecycleOnStopEvent struct {
+	Caller string
 }
 
-func (e Entry) Write(logger Logger) {
-	logger.Log(e)
+// ApplyOptionsError is emitted whenever there is an error applying options.
+type ApplyOptionsError struct {
+	Err error
 }
 
-// Err is a helper for error Fields.
-func Err(value error) Field {
-	return F("error", value)
+// SupplyEvent is emitted whenever a Provide was called with a constructor provided
+// by fx.Supply.
+type SupplyEvent struct {
+	Constructor interface{}
 }
 
-// F is a constructor for a Field.
-func F(key string, value interface{}) Field {
-	return Field{
-		Key:   key,
-		Value: value,
-	}
+// ProvideEvent is emitted whenever Provide was called and is not provided by fx.Supply.
+type ProvideEvent struct {
+	Constructor interface{}
 }
 
-// Field defines a field used inside an internal logging Entry.
-type Field struct {
-	Key   string
-	Value interface{}
+// InvokeEvent is emitted whenever a function is invoked.
+type InvokeEvent struct {
+	Function interface{}
 }
 
-// Logger defines interface used for logging.
-type Logger interface {
-	Log(entry Entry)
+// InvokeFailedEvent is emitted when fx.Invoke has failed.
+type InvokeFailedEvent struct {
+	Function interface{}
+	Err      error
+	Stack    fxreflect.Stack
 }
 
+// StartFailureError is emitted right before exiting after failing to start.
+type StartFailureError struct{ Err error }
+
+// StopSignalEvent is emitted whenever application receives a signal after
+// starting the application.
+type StopSignalEvent struct{ Signal string }
```

Comment:
We should name the Signal "SignalName" or include the `os.Signal` here.
In the code where we supply this, we currently do:

    sig := (<-ch).String()

Instead, if  we do the following, we'll have the os.Signal that fired.

    sig := <-ch